add sourceTimeStamp to ScaleUpdateDetails

yakagami opened the pull request:
This PR adds the ability to get the `sourceTimeStamp` from `ScaleUpdateDetails` in a `GestureScaleUpdateCallback` like so:

```dart
onScaleUpdate: (ScaleUpdateDetails details){
  print(details.sourceTimeStamp);
}
```

`sourceTimeStamp` is necessary when tracking velocity eg.

```dart
VelocityTracker tracker = VelocityTracker.withKind(PointerDeviceKind.touch);
///...
onScaleUpdate: (ScaleUpdateDetails details){
  tracker.addPosition(details.sourceTimeStamp!, details.focalPoint);
}
```

The docs say:

>Having both a pan gesture recognizer and a scale gesture recognizer is redundant; scale is a superset of pan. Just use the scale gesture recognizer.

Currently this is not entirely accurate, and should be fixed, as noted in https://github.com/flutter/flutter/issues/43833#issuecomment-548133779. This PR does not add `sourceTimeStamp` to `ScaleStartDetails` because it is more involved. Specifically, `ScaleStartDetails` can be created in `acceptGesture` which does not have access to the `PointerEvent` to get the `event.timeStamp` (https://github.com/flutter/flutter/blob/54fa25543243e3bf31af6af0c1fef6adabc1d5c1/packages/flutter/lib/src/gestures/scale.dart#L730C5-L730C5).

fixes https://github.com/flutter/flutter/issues/135873. See also https://github.com/flutter/flutter/issues/43833 which added delta and https://github.com/flutter/flutter/issues/49025 which added `numPointers` to `ScaleUpdateDetails` for the reason given above. `sourceTimeStamp` should probably be added to `ScaleStartDetails` as well because it exists in `DragStartDetails` and therefore in `onPanStart`.

I am not sure how to add tests for this, any input about this PR would be appreciated.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

yakagami committed:
add sourceTimeStamp to ScaleUpdateDetails

yakagami committed:
add sourceTimeStamp to ScaleUpdateDetails

moffatman commented:
Start of testing...

```dart
diff --git a/packages/flutter/test/gestures/scale_test.dart b/packages/flutter/test/gestures/scale_test.dart
index fd4bcfaf00..1b5429b38a 100644
--- a/packages/flutter/test/gestures/scale_test.dart
+++ b/packages/flutter/test/gestures/scale_test.dart
@@ -27,12 +27,14 @@ void main() {
     double? updatedHorizontalScale;
     double? updatedVerticalScale;
     Offset? updatedDelta;
+    Duration? updatedSourceTimestamp;
     scale.onUpdate = (ScaleUpdateDetails details) {
       updatedScale = details.scale;
       updatedHorizontalScale = details.horizontalScale;
       updatedVerticalScale = details.verticalScale;
       updatedFocalPoint = details.focalPoint;
       updatedDelta = details.focalPointDelta;
+      updatedSourceTimestamp = details.sourceTimeStamp;
     };
 
     bool didEndScale = false;
@@ -56,6 +58,7 @@ void main() {
     expect(updatedScale, isNull);
     expect(updatedFocalPoint, isNull);
     expect(updatedDelta, isNull);
+    expect(updatedSourceTimestamp, isNull);
     expect(didEndScale, isFalse);
     expect(didTap, isFalse);
 
@@ -65,10 +68,12 @@ void main() {
     expect(updatedScale, isNull);
     expect(updatedFocalPoint, isNull);
     expect(updatedDelta, isNull);
+    expect(updatedSourceTimestamp, isNull);
     expect(didEndScale, isFalse);
     expect(didTap, isFalse);
 
-    tester.route(pointer1.move(const Offset(20.0, 30.0)));
+    const ts1 = Duration(milliseconds: 10);
+    tester.route(pointer1.move(const Offset(20.0, 30.0), timeStamp: ts1));
     expect(didStartScale, isTrue);
     didStartScale = false;
     expect(updatedFocalPoint, const Offset(20.0, 30.0));
@@ -77,6 +82,8 @@ void main() {
     updatedScale = null;
     expect(updatedDelta, const Offset(20.0, 30.0));
     updatedDelta = null;
+    expect(updatedSourceTimestamp, ts1);
+    updatedSourceTimestamp = null;
     expect(didEndScale, isFalse);
     expect(didTap, isFalse);
     expect(scale.pointerCount, 1);
```

yakagami committed:
add tests

yakagami committed:
add tests

yakagami commented:
I've added some tests. In line with https://github.com/flutter/flutter/blob/master/packages/flutter/test/gestures/drag_test.dart I've started the first event at `const Duration(milliseconds: 100)` and increased each event by 100 milliseconds.

goderbauer commented:
(triage): Looks like a lot of checks are failing. Can you take a look and fix those issues?

yakagami committed:
fix tests

yakagami committed:
Merge branch 'master' of https://github.com/yakagami/flutter

yakagami committed:
fix tests

yakagami committed:
fix tests

yakagami commented:
@goderbauer thanks. All checks should be passing now.

yakagami committed:
removed extra declaration casued by merge conflict

yakagami committed:
Merge branch 'master' into master

Renzo-Olivares commented:
Hi @yakagami, thank you for your contribution! This seems like a reasonable addition and I agree with the points made in your PR description. I do think we should add this to `ScaleStartDetails` as well and am happy to provide an explanation on how to do that.

It looks like there are two potential starts to a scale gesture. One is `addAllowedPointerPanZoom` which is called when a scale start without a `PointerDownEvent` (the user first touching down on the screen). The second is `addAllowedPointer` which is called when a scale starts when the user first touches down on the screen and then pinches to scale. Given these two cases I think we should track the initial event time stamp in `Duration? _initialEventTimestamp;` and set this variable in `addAllowedPointer` and `addAllowedPointerPanZoom`. We can then pass `_initialEventTimestamp` to `ScaleStartDetails` which is instantiated in `_dispatchOnStartCallbackIfNeeded`. We should also make sure to `null` the `_intialEventTimestamp` after it is used directly in `_dispatchOnStartCallbackIfNeeded`. Let me know if this explanation makes sense or if it needs some clarifying!

yakagami committed:
add sourceTimeStamp to ScaleStartDetails

yakagami committed:
Merge branch 'master' of https://github.com/yakagami/flutter

yakagami committed:
make _initialEventTimestamp null after it's used

Renzo-Olivares commented on packages/flutter/lib/src/gestures/scale.dart:
```diff
@@ -730,6 +740,7 @@ class ScaleGestureRecognizer extends OneSequenceGestureRecognizer {
+          sourceTimeStamp: _initialEventTimestamp,
         ));
       });
     }
```
nit: should probably set `_initialEventTimestamp` to `null` at the end of this method.

yakagami replied:
woops, committed too fast. Should be fixed now.

yakagami commented:
@Renzo-Olivares How should I test for this? Do I add the time stamp to `tester.route(down)` or is this already covered in the first pointer move? (`tester.route(pointer1.move(const Offset(20.0, 30.0), timeStamp: const Duration(milliseconds: 10)));`)

yakagami commented:
hmm testing this shows that `_initialEventTimestamp` is always null when called in `_dispatchOnStartCallbackIfNeeded`... perhaps there are more than two potential starts to a scale gesture?

yakagami committed:
always make _initialEventTimestamp null after use.

yakagami committed:
make sure _initialEventTimestamp is set on first interaction

yakagami committed:
remove print statements

yakagami committed:
test that initialSourceTimestamp isNull after every non-starting move.

yakagami committed:
nit: move declaration to bottom

yakagami commented:
let me know what you think of my solution to the above problem. I added `_initialEventTimestamp = event.timeStamp;` to the `handleEvent` method, namely in `if (event is PointerMoveEvent){` and `else if (event is PointerPanZoomStartEvent) {` which allows the tests to pass. I'm a bit worried about this though because I don't know _why_ this works, only that it does. I added tests to make sure that any event that is not expected to emit a `ScaleUpadateDetails` callback does not do so. I also tested that  `_initialEventTimestamp` is null when `_advanceStateMachine` emits the `ScaleUpdateDetails` callback. I observe that the `if (event is PointerMoveEvent) {` branch of `handleEvent` seems to be the start of a gesture as it contains the line `shouldStartIfAccepted = true;`. But if this is the case then perhaps `_initialEventTimestamp = event.timeStamp;`  is not needed in the other cases? I left them in for now.

Renzo-Olivares commented:
> hmm testing this shows that _initialEventTimestamp is always null when called in _dispatchOnStartCallbackIfNeeded... perhaps there are more than two potential starts to a scale gesture?

Do you mean testing on device or writing a test? I just tested on my macOS device and i'm able to get a non-null `timeStamp` in `addAllowedPointer` and `addAllowedPointerPanZoom` and `dispatchOnStartCallbackIfNeeded`. Let me know if you are experiencing something different on a different platform.

yakagami commented:
@Renzo-Olivares It was null in the unit test. I didn't try it on any devices. eg:

```dart
    Duration? initialSourceTimestamp;
    scale.onStart = (ScaleStartDetails details) {
      didStartScale = true;
      updatedFocalPoint = details.focalPoint;
      initialSourceTimestamp = details.sourceTimeStamp;
    };
    // later
    tester.route(pointer1.move(const Offset(20.0, 30.0), timeStamp: const Duration(milliseconds: 10)));
    expect(initialSourceTimestamp, const Duration(milliseconds: 10)); //fails
```

Renzo-Olivares commented:
Hi @yakagami I investigated the testing issue a little and found that the `PointerDownEvent`s do not have a `timeStamp` defined in the tests. The `PointerDownEvent` is where we read the `_initialEventTimestamp` from and pass it to `ScaleStartDetails`. `PointerMoveEvent` in the tests is given a `timeStamp` which is why tracking `_initialEventTimestamp` worked when done in `handleEvent` since that is where move events are handled. 

I think we should not track `_initialEventTimestamp` in `handleEvent` and we should provide the `timeStamp` to `PointerDownEvent`s in tests to fix the issue you were experiencing. Let me know if this makes sense.

yakagami committed:
don't track start timestamp in handleEvent

yakagami commented:
@Renzo-Olivares Thanks for pointing that out. I was confused because In the tests, we have 
```dart
final PointerDownEvent down = pointer1.down(Offset.zero);
expect(didStartScale, isFalse);
//later
tester.route(pointer1.move(const Offset(20.0, 30.0)));
expect(didStartScale, isTrue);
```

I initially read this code as implying that the event does not start until the first move.

After changing it back, however, tests fails in the two cases cases where a scale starts without an accompanying `PointerDownEvent`. eg

```dart
    //line 277
    // Continue panning with one finger
    tester.route(pointer3.move(Offset.zero, timeStamp: const Duration(milliseconds: 150)));
    expect(didStartScale, isTrue);
```
Here a scale start occurs without a `pointerDownEvent`, and  initialSourceTimestamp should be `Duration(milliseconds: 150)` but it is null.

yakagami committed:
Merge branch 'master' into master

Renzo-Olivares commented:
Hi @yakagami, apologies for the delay. I think the issue is that `_dispatchOnStartCallbackIfNeeded` is called in two places by `ScaleGestureRecognizer`, [here](https://github.com/flutter/flutter/blob/db7ef5bf9f59442b0e200a90587e8fa5e0c6336a/packages/flutter/lib/src/gestures/scale.dart#L692-L695) and [here](https://github.com/flutter/flutter/blob/db7ef5bf9f59442b0e200a90587e8fa5e0c6336a/packages/flutter/lib/src/gestures/scale.dart#L730-L734).

When the latter is called before the former it nulls the `_initialEventTimestamp` and then the other call will not get a chance to have that timestamp. We should save the more accurate timestamp when that happens like below.

```dart
  if (_state == _ScaleState.accepted && shouldStartIfAccepted) {
    _initialEventTimestamp = event.timeStamp;
    _state = _ScaleState.started;
    _dispatchOnStartCallbackIfNeeded();
  }
```
This will also affect lines 124/240 of the test. You should change the duration to match that of the `updatedSourceTimestamp` above these lines.

This should get all the tests passing. Let me know if that helps.

yakagami committed:
Merge branch 'flutter:master' into master

yakagami committed:
make sure _initialEventTimestamp is set when a scale starts without a PointerDownEvent

yakagami committed:
Merge branch 'master' of https://github.com/yakagami/flutter

yakagami commented:
@Renzo-Olivares thanks for your comment, the test are now passing. Do you suggest I move this out of the "main" scale test and into a seperate `testGesture`? Something like
```dart
testGesture("scale gesture should emit ScaleStartDetails and ScaleUpdateDetails callbacks"
```

Renzo-Olivares commented:
Glad to hear that worked @yakagami. That sounds like a good change, maybe change the test name so it include a mention of timestamps/ reporting timestamps.

yakagami committed:
move timestamp tests to new subtest

yakagami committed:
Merge branch 'master' into master

yakagami commented:
@Renzo-Olivares done.

Renzo-Olivares commented on packages/flutter/test/gestures/scale_test.dart:
```diff
@@ -1356,4 +1356,296 @@ void main() {
+    expect(didTap, isFalse);
+    expect(initialSourceTimestamp, isNull);
+
+    // One-finger panning
```
consider ending these comments with a period, here and in other places.

yakagami replied:
These are copied from the origional tests. Should those be changed as well?

Renzo-Olivares replied:
Not necessary for this change, but would be welcome.

yakagami replied:
done.

Renzo-Olivares commented on packages/flutter/test/gestures/scale_test.dart:
```diff
@@ -1356,4 +1356,296 @@ void main() {
     scale.dispose();
   });
+
+  testGesture('ScaleStartDetails and ScaleUpdateDetails callbacks should contain thier event.timestamp', (GestureTester tester) {
```
`thier` -> `their`

yakagami committed:
fix spelling of "their"

yakagami committed:
Merge branch 'master' of https://github.com/yakagami/flutter

yakagami committed:
Merge branch 'master' into master

yakagami committed:
Merge branch 'master' into master

yakagami committed:
add periods to comments

yakagami committed:
Merge branch 'master' of https://github.com/yakagami/flutter

Renzo-Olivares commented on packages/flutter/lib/src/gestures/scale.dart:
```diff
@@ -129,6 +130,12 @@ class ScaleStartDetails {
   /// recognizer.
   final int pointerCount;
 
+  /// Recorded timestamp of the source pointer event that triggered the drag
```
nit: consider changing `drag` to `scale`.

Renzo-Olivares commented on packages/flutter/lib/src/gestures/scale.dart:
```diff
@@ -225,6 +233,12 @@ class ScaleUpdateDetails {
   /// recognizer.
   final int pointerCount;
 
+  /// Recorded timestamp of the source pointer event that triggered the drag
```
nit: consider changing `drag` to `scale`.

Renzo-Olivares reviewed (commented):
Hi @yakagami, just had a few last comments. Also try rebasing your pr to `master` and see if that fixes the failing customer tests. They seem unrelated to this pr.

yakagami committed:
fix comment to reflect gesture type (drag -> scale)

yakagami committed:
Merge branch 'master' into master

Renzo-Olivares reviewed (approved):
LGTM

yakagami committed:
Merge branch 'master' into master

yakagami committed:
Merge branch 'master' into master

gspencergoog reviewed (approved):
![32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d](https://user-images.githubusercontent.com/8867023/37737237-66e9c9dc-2d10-11e8-8797-13a405eace49.png)
